### PR TITLE
SDIT-897 Attendance sync multiple ended allocations

### PR DESCRIPTION
**Pull request by mikehalmamoj**
We made an incorrect assumption that when we are looking for an attendance record to sync. there will be exactly one active allocation. This is not always true - we could receive a request to expire an attendance *after* the prisoner has been deallocated from the course. Therefore we just get the latest allocation rather than the active allocation.

### CODE REVIEW

File: src/main/resources/migrations/common/V2023.07.04.2__fix_attendance_sync_view.sql [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+       a.comment,
+       a.status,
+       a.pay_amount,
+       a.bonus_amount,
+       a.issue_payment,
+       a2.start_date as allocation_start_date
+from attendance a
+         join scheduled_instance si on a.scheduled_instance_id = si.scheduled_instance_id
+         join allocation a2 on si.activity_schedule_id = a2.activity_schedule_id and
+                               a.prisoner_number = a2.prisoner_number
```
Comment (mikehalmamoj) [inline src/main/resources/migrations/common/V2023.07.04.2__fix_attendance_sync_view.sql:24]: The change to the SQL is that we used to also look for `a2.prisoner_status != 'ENDED'` when joining on allocations - now we just get them all.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SynchronisationService.kt [hunk 2]
```diff
@@ -1,14 +1,13 @@
...
 @Service
 class SynchronisationService(private val repository: AttendanceSyncRepository) {
-  fun getAttendanceSync(attendanceId: Long): AttendanceSync =
-    repository.findById(attendanceId)
-      .orElseThrow { EntityNotFoundException("Attendance sync not found: $attendanceId") }
-      .toModel()
+  fun findAttendanceSync(attendanceId: Long): AttendanceSync? =
+    repository.findAllByAttendanceId(attendanceId)
+      .maxByOrNull { it.allocationStartDate }
+      ?.toModel()
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SynchronisationService.kt:12]: As we now get all allocations we are only interested in the latest one.

File: src/test/resources/test_data/seed-activity-id-17.sql [hunk 3]
```diff
@@ -38,7 +38,7 @@ insert into allocation(allocation_id, activity_schedule_id, prisoner_number, boo
 values (5, 2, 'A22222A', 10002, 3, '2022-10-10', null, '2022-10-10 10:00:00', 'MRS BLOGS', null, null, null, null, null, null, 'ACTIVE');
 
 insert into allocation(allocation_id, activity_schedule_id, prisoner_number, booking_id, prison_pay_band_id, start_date, end_date, allocated_time, allocated_by, deallocated_time, deallocated_by, deallocated_reason, suspended_time, suspended_by, suspended_reason, prisoner_status)
-values (6, 1, 'A33333A', 10003, 2, '2022-10-20', null, '2022-10-20 09:00:00', 'MRS BLOGS', null, null, null, null, null, null, 'SUSPENDED');
+values (6, 1, 'A33333A', 10003, 2, '2022-10-20', null, '2022-10-20 09:00:00', 'MRS BLOGS', null, null, null, null, null, null, 'ENDED');
```
Comment (mikehalmamoj) [inline src/test/resources/test_data/seed-activity-id-17.sql:41]: Test data changed to have multiple allocations where none is currently active - but we're only interested in the latest one.

File: src/main/resources/migrations/common/V2023.07.04.2__fix_attendance_sync_view.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
+-- =============================================
+-- ATTENDANCE SYNC VIEW
+-- =============================================
+
+CREATE OR REPLACE VIEW v_attendance_sync AS
```
Comment (matt-crowson-moj) [inline src/main/resources/migrations/common/V2023.07.04.2__fix_attendance_sync_view.sql:5]: So @leejacobson-moj also has some migrations and I think this clashes.

So perhaps check with him first - this is the [PR](https://github.com/ministryofjustice/hmpps-activities-management-api/pull/441)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SynchronisationService.kt (hunk 2 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SynchronisationService.kt:12]: So one attendance row has a specific scheduled_instance as its parent, and that has a specific activity_schedule as its parent, but there could potentially be a few allocations of the same prisoner to that activity_schedule (if they've been allocated, deallocated and and allocated again)?  So this change will always take the latest allocation. Sounds reasonable.

**Review by timharrison-moj — DISMISSED** [review]
LGTM

**Review by matt-crowson-moj — APPROVED** [review]
lgtm

**Pull request merged**